### ISSUE TITLE
Make the $ operator in GDScript cache node references

### ISSUE THREAD

**Issue by aaronfranke** [godotengine/godot-proposals#996]
**Describe the project you are working on:** https://github.com/godotengine/godot-demo-projects/ but this applies to any project made in GDScript.

**Describe the problem or limitation you are having in your project:**

Here is a code snippet from Dodge the Creeps:

![](https://user-images.githubusercontent.com/1646875/76044779-4e4a2000-5f29-11ea-96c3-161619caa301.png)

The problem is that this code looks elegant, but it isn't very performant. Every time this method is called, it has to call `get_node()` 6 times, and 2 of those times are getting the same node (HUD). We can reduce this from 6 to 5 by caching the HUD reference like this:

![76044779-4e4a2000-5f29-11ea-96c3-161619caa301](https://user-images.githubusercontent.com/1646875/83560643-8ccd8780-a4e4-11ea-8af3-8bd40635e8bc.png)

However, this method is still very inefficient, since it calls `get_node()` 5 times every time the method is ran. The best simple option that I can see is to use `onready var` like this:

![76044779-4e4a2000-5f29-11ea-96c3-161619caa301](https://user-images.githubusercontent.com/1646875/83561043-2b59e880-a4e5-11ea-8b89-fe884dec06af.png)

This means that `get_node()` is only called 5 times once when this node is created, and this method does not have to call `get_node()` at all when the method is ran. However, now this code is ugly, since it is much longer than it could be and we've lost the green syntax highlighting in `new_game()` that tells us we're working with node children.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

The proposal is to make the `$` operator cache references on ready, so that user code looks like the top image, but behaves like the bottom image.

* For projects currently caching references via `onready var`, this would make code more elegant. 

* For projects currently using `$` everywhere, this would increase performance.

In cases where `get_node()` needs to be called every time, such as if nodes are created and deleted, users can just use `get_node()` instead of `$`.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

See above for examples of user code, but @vnen would be the one handling the implementation.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

It can be worked around, but this is something that will be used often.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Yes, because it will be used often.

**Comment by nathanfranke** [comment]
A few notes/questions:

1. This would technically break compatibility so any PR should definitely note that.
2. By how much does the performance improve with this workflow?
3. This actually makes a lot of sense since it is extremely unlikely to do this with dynamically added nodes like `$Node@@2`. The `$` should be used for accessing nodes that are already in the scene before ready.

**Comment by groud** [comment]
I don't think it is a good idea. Some users might rely on dynamically changing sub-nodes, so this behavior would lead to undefined references.

The thing is that I am not even sure this would lead to a significant performance improvement for several reasons:
1) internally the get_node functions uses NodePath (a list of StringName) instead of Strings, so everything is using reference and the string is not re-analyzed everytime,
2) accessing the node following the path is a very fast operation, which should not be long to execute at all. Most of the time it's a matter of iterating over the few components the path contains
3) adding a cache system requires checking when the cached values becomes invalid, as we don't want the engine to crash if a reference becomes wrong. This requires additional processing which might not be worth the time gained by caching the node.

Anyway, before trying to improve performance in a given part of the code, we have to make sure this is a bottleneck for some users, or that the gain will be really significant for everyone. So unless someone make a benchmark proving that this part of the code is slow, it's not worth making the engine code significantly more complex.

**Comment by Zireael07** [comment]
I store node references in vars and it does make a performance difference, at least in functions that are running often (e.g. physics_process and process() ).
I don't have a benchmark on hand.

**Comment by groud** [comment]
> I store node references in vars and it does make a performance difference

If you access the node several times per frame it surely likely does. But I believe such case is an optimization to be done on the user's side, as he can know if the node is eventually going to be replaced or removed from the hierarchy during the processing.
I believe the most common usage is to access a node once per frame, so in such situation I am not sure is it makes a big difference

**Comment by vnen** [comment]
I feel this is too much magic. If the script is long, you're now adding a lot of stuff on the user's `_ready()` function behind their back. While it's hard to make it too slow, there is an added penalty. Maybe the user don't have a `_ready()` function at all and will be wondering why the profiler says there is some time spent in `_ready()`.

And the more obvious: what do you do when a node is freed? Because if you call `get_node()` you'll get `null` but if you are using a cache then you just have pointer to an invalid object.

Also, the user might remove a node and replace with another using the same name. If you keep a cache, then the `$Node` shorthand will only point to the old node.

It's also possible that you make some cases worse. If the user is using `_ready()` to connect signals:

```gdscript
func _ready():
    $Button.button_up.connect(_on_Button_button_up)
```

Now it has a cached value that will never be used again.

Of course we can be clever about usage, but I do think this will lead to bugs into the user code that will be hard to pinpoint. The only way to get around those would be using `get_node()` instead  of `$`. What would happen is that the shorthand would get bad rap and people would stop using it, making this optimization pointless.

----

If we are going to be clever about usage, then we can consider adding warnings on cases where this is used inside functions called often. Like this:

```gdscript
func _process(delta):
    $Sprite2D.rotate(PI * delta) # Warning: Consider caching the reference to "$Sprite2D" to avoid getting the node from tree every frame.
```

**Comment by Zireael07** [comment]
Hm, good point. Could we get a caching equivalent of get_node() (cache_get_node() or some such), or would that require a new shorthand?

**Comment by aaronfranke** [comment]
> And the more obvious: what do you do when a node is freed? 

As stated, if you are working with a non-static node tree you would just use `get_node()` instead of `$`. I really don't see the problem, since in most cases I've seen, `$` is used to access nodes that don't change, and you can just use `get_node()` otherwise.

**Comment by groud** [comment]
> As stated, if you are working with a non-static node tree you would just use get_node() instead of $.

You are talking about the user point of view, but the problem is not there at all.

What is difficult to do is implementing the caching system while guaranteeing that the reference is reset to null when the node is replaced, moved or removed from the tree. This would require adding checks in every move or delete operation to make sure that we cannot have a wrong pointer somewhere in the code (as the engine should not crash at all). This will, as a consequence, slow down those other operations.

**Comment by aaronfranke** [comment]
If a node is able to be replaced, moved, or removed, users should use `get_node()` instead of `$`.

**Comment by groud** [comment]
I think you don't get the point. This does not changes anything. Even if the user know that, the engine should not crash at all. And this is not negotiable.

Even if the user is not supposed to move or remove the node because "it's not supposed to work", we still have to add checks to simply avoid crashes. Checks that are going to slow down everything else.

**Comment by aaronfranke** [comment]
If that is true, wouldn't crashes also exist with `onready var x = $X`? I haven't noticed any.

**Comment by groud** [comment]
Ah yes you are right, I did not think about that. I am not sure about how this is handled in the engine code, it's true that variable referencing nodes should not cause the engine crash too.

**Comment by nathanfranke** [comment]
```
onready var x = $X
```
would translate to
```
onready var internal_x = get_node("X")
onready var x = internal_x
```

**Comment by aaronfranke** [comment]
@nathanfranke I meant `onready var x = $X` with the current behavior.

**Comment by Zylann** [comment]
I don't really like this either, notably because caching is only valid if you know your branch will be static. Now from my experience this is very often the case, and I use `onready` caching pretty much all the time (which is why at first it sounds like a good idea). However I'm not only doing it for performance: in my mindset, references to external stuff (assets, directories, other scripts, node paths) must appear on top of the script, and only once, instead of being spread throughout the script. It also makes it easier to re-arrange the tree or rename nodes.

> onready var x = $X

>would translate to

>onready var internal_x = get_node("X")
>onready var x = internal_x

Besides, I wouldnt want this to happen, it would make my use case worse. Other than that, I'm not the target audience :p

**Comment by Mantissa-23** [comment]
Here's a thought:

`var x = $X` - Maintain current behaviour so as to not break compatibility

`var x = $$X` - Cache variable implicitly. Any time `$$X` is accessed again, the cached reference will be used.

This doesn't break compatibility, adds a shorthand that both new and experienced developers can take advantage of, and still allows for developers to explicitly decide whether or not they want a node cached.

**Comment by nathanfranke** [comment]
My proposal which also endorses better code readability with multiple scripts and nodes

```swift
# shorthand for onready var label = $Label
some_keyword label, "Label"

func _process(delta):
    label.text = "FPS: " + str(1 / delta)
    
    # something else you could do
    other_node.animation_player.play()
```

**Comment by aaronfranke** [comment]
I still stand by my original proposal. For any given edge case where the proposed behavior of `$` is not desired, you can just use `get_node()` instead. We can break compat in Godot 4.0.

**Comment by Zylann** [comment]
@aaronfranke I want to use `$` because it makes query faster (better loading and instancing times), but I don't want it to dupe all my `onready` vars in which I already use it.

**Comment by vnen** [comment]
> My proposal which also endorses better code readability with multiple scripts and nodes
> 
> ```python
> # shorthand for onready var label = $Label
> some_keyword label, "Label"
> ```
That's not much different of what you can do today:

```gdscript
@onready var label = $Label
```

**Comment by BeayemX** [comment]
I don't see the problem with storing everything in variables. I always store nodes in variables in `onready`. In my opinion it makes the code cleaner. You see all node-dependencies in one place.
And if the hierarchy changes you have one place where you can adjust everything.

When using the `$` throughout your code, you have to find every instance where you used it and change it, which is quite error prone.

**Comment by aaronfranke** [comment]
@BeayemX You can just do `onready var x = get_node("X")` in your case with this proposal.

**Comment by Zylann** [comment]
@aaronfranke and that's precisely not good, because `get_node()` has to build a nodepath to do this, while `$` makes it at compilation time, in addition to be shorter to write.

**Comment by aaronfranke** [comment]
@Zylann Interesting, I didn't know that. You can just do `const x = NodePath("test")` to make a compile-time NodePath. The use case you are describing can similarly be worked around with lines at the top of scripts, but personally I have never had a need to have a compile-time NodePath with a dynamic node, so I think it's a less common case.

In the end we both want better performance. Storing references to non-changing nodes is arguably the most common case, which is why I want `$` to do that. I don't want to have many lines at the top of my script, and lose the special green syntax highlighting, if I can avoid it. I want the most common operations to be high-performance and elegant.

**Comment by Zylann** [comment]
> You can just do const x = NodePath("X") to make a compile-time NodePath

And write even more code, sorry^^"

> I have never had a need to have a compile-time NodePath with a dynamic node

The case I described isn't dynamic, and I use it in basically *every* script that accesses child nodes, in all my projects (I'm not the only one doing that). Also demo projects aren't representative of that.

> In the end we both want better performance

In my case (and some others) this is not just about performance. I also want more maintainable code through decoupling, which happens to be done by explicit caching for fixed nodes, and `const` paths for dynamic ones (very rare, tho, I give you that). Which means implicit caching gets in the way by either duping vars needlessly, or forcing to downgrade my code and write more lines than before. The idea of auto-caching `$` comes from good intention, but I can't agree to have it because of that use case conflict. That's encouraging an existing pattern by stepping over another existing pattern.
Maybe if you can guarantee member vars assigned from `$` won't get a dupe, that would solve it, thought it's exceptional behavior which you could only guarantee if these vars were marked read-only.

**Comment by aaronfranke** [comment]
> Maybe if you can guarantee member vars assigned from `$` won't get a dupe

That should be possible to implement if the only usage is in an `onready var`, but I think that `$` would still have to make a duplicate variable if there is also other uses of it elsewhere in the script. This should be implementable without any look-ahead if done right.

Also, if removing the existing behavior of `$` is deemed unacceptable, I would be fine with @Mantissa-23 's suggestion of `$$`, but do note that I would end up using `$$` basically everywhere and it does seem a bit weird to make the more common operation longer in characters.

**Comment by Mantissa-23** [comment]
I agree it's a bit weird but maintaining backwards compatibility is important. Breaking backwards compatibility too frequently can bleed users pretty easily. Breaking compatibility like that might make sense for Godot 4.0, but I also think a lot of people are expecting 4.0 to be largely a "backend" upgrade that just makes their game better without really affecting the way they make it.

**Comment by PetePete1984** [comment]
I vaguely recall `var x = @"/some/node"` being usable in place of `var x = NodePath("/some/node")`, so maybe `onready var x = get_node(@"/some/node")` is the secret sauce?

**Comment by nathanfranke** [comment]
I wonder if it would be a better practice to encourage users to use export nodes (once they are supported) rather than hard-coding node names.

```swift
export(Label) var my_label

func _process(delta):
    my_label.text = "Bad FPS Estimation: " + str(1/delta)
```

**Comment by aaronfranke** [comment]
@PetePete1984 Based on [the documentation](https://docs.godotengine.org/en/latest/classes/class_nodepath.html) it does seem like `get_node(@"X")` is the same as `$X`

**Comment by aaronfranke** [comment]
Another advantage of this proposal is that it means `$` will error immediately if the node doesn't exist. I ran into a situation today where I was refactoring and `$` had a name mismatch, and I didn't notice the problem for a few minutes until that line of code was called.

**Comment by Zylann** [comment]
@aaronfranke I still stand on my opinion. This proposal is 100% beneficial only for people coding without any `onready` declarations, the downside for others being a downgrade of what's needed to write or a memory dupe.

> export(Label) var my_label

This being a killer feature, better than both options IMO.

**Comment by fabriceci** [comment]
> It's also possible that you make some cases worse. If the user is using _ready() to connect signals:

@vnen Is it possible to simply not make this "caching" happening in the `_ready` and `_init` (for optimisations purpose)? They are the only functions where this proposal is not relevant.

> I feel this is too much magic.

`$Node` is an alias of `get_node("Node")`, the proposal make `$Node` an alias of `onready var _node =  $Node`. Nearly the same magic, no new behaviour, just a different alias that make possible to hide big bunch of useless variables.

> the user might remove a node and replace with another using the same name. If you keep a cache, then the $Node shorthand will only point to the old node.

Seems to be an edge case that I never saw. I think is a bad practice to use $ on nodes that can be replaced. If really needed, the hidden array that store cached variables can be made accessible by functions to be able to make update/delete operations.

> @aaronfranke I still stand on my opinion. This proposal is 100% beneficial only for people coding without `onready` declarations

I disagree this proposal will benefit to all users, and a lot. They are main benefits:

- Clearer for users: $ for static node, get_node() for dynamic ones

- Save time by avoiding a repetitive task (and boring)

- Optimise the code for those (often beginners) who use $ alias in process loop. 

- Prevent error in the editor, if the node is static, it should exist in the tree.

- And the best: to make the script **a way shorter/readable** by removing endless useless declaration. **They are useless because they are only added for caching/performance purpose, they are not "real" variables for the script**. On my "serious" amateur projects, I have often scripts where when you open them in the editor, you only see a whole page of `onready var` declarations if you don't scroll. That makes scripts **hard to read and understand**, and this situation is probably worse on professional projects.

The only drawback I see, is to understand that this alias should be used for static node (which already represents the vast majority of usage), the price seems cheap.

**Comment by Zylann** [comment]
> They are useless because they are only added for caching/performance purpose, they are not "real" variables for the script.

Did you read my previous posts?
They are far from useless. This kind of practice tells coders right off the bat which nodes the script depends on. It makes it easier to maintain when node names or their path change as the project evolves. And they can be typed, annotated, and commented, without requiring Godot and the scene to be open. Eventually exporting node types directly could even get rid of the need to hardcode them. We should not assume everyone should use `$` all over the place. There are very good reasons not to.

**Comment by fabriceci** [comment]
> Did you read my previous posts?

Yes I'm not convinced by them actually, but I learned the @"" syntax in the process :)

> This kind of practice tells coders right off the bat which nodes the script depends on. It makes it easier to maintain when node names or their path change as the project evolves.

When I see the screenshots of @aaronfranke in the first post. I think the opposite, the version without extra var is more readable (+ extra highlight), I wouldn't argue further why these variables are unnecessary because at the end, we can agree that's a matter of preference.

But "your" preference/practice require to write many extra lines of codes that some (I suspect a lot) thinks they are useless. Moreover, this proposal doesn't prevent your practice from being applied **if you want**:

```python
var node = $node
```

> And they can be type-hinted, annotated, and commented, without requiring Godot and the scene to be open:

Same, nothing prevent you to store and add a type **if you want**:

```python
var node: MyNode = $node
```

I don't understand your opposition if nothing in this proposal prevent you to apply your preference/best practice.

For the other arguments:
  - breaking changes: we are in the scope of Godot 4.0 with GD 2.0, so it's irrelevant, especially since the majority of users will not see any difference.
  - for the exotic case when you use `$`with dynamic nodes, I never saw that, but this proposal require you will need the change to `get_node(@"X")`.

**Comment by reduz** [comment]
You guys know how I think about these things, I think that most of the time, _explicitness is better_, and things that hide what happens under the hood to the programmer in a non-obvious way, even if it means writing a bit less code, is generally bad.

Originally, when I created the $ syntax, I saw that this may be a problem in cases where you wanted performance. Still, by experience, in far most cases where you use this, you don't really need that performance. It's only in a few situations that you may need the node to access it faster. 

For this, I thought I could add a "cached $Node" syntax to tell the programmer that this node is cached, but the problem with this is that _when_ this node is cached still remains a mystery. This is why the **onready** keyword was born. Its a much simpler and flexible solution, and very explicit in nature, so its obvious what it does when you look at it. I don't think this needs to be changed.

**Comment by reduz** [comment]
Still, an alternative that I always thought could work may be creating something like:

```
onready $Node
```

This way, if you already have code written and you want to optimize it, its just easier and its still fairly explicit.

**Comment by aaronfranke** [comment]
Here is a small project I made to benchmark the time required to get nodes with `$`: [BenchmarkingDollarSign.zip](https://github.com/godotengine/godot-proposals/files/5449986/BenchmarkingDollarSign.zip)

On my system, I get output similar to this (the last part is a fairly trivial operation to give a sense of scale):

```
Average time to get nodes on process: 116.389646 microseconds per frame
Total time wasted getting all nodes on process: 0.640725 seconds
Total time wasted getting each node on process: 0.006407 seconds
Average time to set pause mode: 51.663034 microseconds per frame
```

It seems that each call to `$` takes about 1 microsecond on average. This means that if you have a project with ~166 calls to `$` per frame, and your project runs at 60 FPS, then an entire 1% of your execution time is dedicated to the node getting logic. On the other hand, if you have a project with 166 `onready var x = $X` lines, you would have 1% extra time per frame compared to if you used `$` everywhere, but then you would have an extra 166 lines of code to maintain.

**Comment by Zylann** [comment]
I just tried that benchmark on a Ryzen 5, here `$` results in process are:
```
Average time to get nodes on process: 46.86875 microseconds per frame
Average time to set pause mode: 22.041667 microseconds per frame
```
But I noticed something interesting: if I execute those 10000 times per process instead of once in the goal of getting more "isolated" result with less printing, I get:
```
Average time to get nodes on process:  237282.727273 microseconds per frame
Average time to set pause mode: 106606 microseconds per frame
```
Which, when divided by 10000 to be comparable to the first result, gives 23.7 microseconds per frame for `$`, and 10.7 microseconds to assign pause mode. It doubled in speed. Not sure what to think of that, although the ratio is similar.

When you say `166` extra lines to maintain, I disagree. Sure you have 166 extra lines [not all in the same script] if you completely disregard what they are, but these lines can *lower the base maintainance cost* and have more possibilities (cf previous comments). That might just be less true if you access a node only once in your script, of course. It's coding preferences anyways.

I still think, as said in https://github.com/godotengine/godot-proposals/issues/996#issuecomment-671316234, that we can have caching, if it does not redundantly occur from code written in onready initializers.

**Comment by HeartoLazor** [comment]
A change I would like to propose to this system is cache the $ calls BEFORE ready time, so when you instance the node, the $ calls are already cached.
With this approach the user can preinstance the levels in a menu thread for example and only process the ready when the game start.
I already did something similar as optimization using NOTIFICATION_INSTANCED.

**Comment by Zylann** [comment]
#1048 is another option which doesn't even need to hardcode the path in a script, on top of being meant to run at instancing time.

**Comment by h0lley** [comment]
alright I just read through the entire thread and while the critique made sense to me, I did not see a single comment arguing that `$$` was a bad idea.

absolutely; let the programmer decide where it makes sense to use the current behavior and where caching would be useful. but that doesn't mean we can't have this.

it seems like there are two camps of us:
those who like to declare a lot of node references as onready, and those of us who prefer skipping this and making use of `$` instead. personally I belong to the second camp as I dislike having to add a property to class scope for every node I am going to use in code (or even two variables when the reference is derived from an exported NodePath). I am also quite fond of the green syntax highlighting communicating clearly that we are dealing with a node. but this shouldn't be about which one of these two patterns is superior, of course. let's support both, please.

by the way, here is an up to date performance test I just ran in the current alpha:

```
# 1 mil iterations each
236ms # direct node reference access
278ms # $NodePath
278ms # get_node(^"LiteralNodePath")
533ms # get_node("NodePathAsString")
```

<details>
<summary>source</summary>

```
extends Node

const ITERATIONS = 1000000


func test(code: Callable):
	var time_before = Time.get_ticks_msec()

	for i in ITERATIONS:
		code.call()

	var time_taken = Time.get_ticks_msec() - time_before

	print(str(time_taken) + "ms - " + str(code.get_method()))


@onready var node = $Node2D

func access_node_reference():
	node.position.x = randf()

func use_get_node_dollar():
	$Node2D.position.x = randf()

func use_get_node_with_str():
	get_node("Node2D").position.x = randf()

func use_get_node_with_np():
	get_node(^"Node2D").position.x = randf()


func _ready():
	print("Running tests with " + str(ITERATIONS) + " iterations each...")

	test(access_node_reference)
	test(use_get_node_dollar)
	test(use_get_node_with_str)
	test(use_get_node_with_np)

	await get_tree().create_timer(0.5).timeout
	get_tree().quit()
```
</details>

as you can see, performance of get_node() with a string passed is poor, however with a constant NodePath (which is always the case when using `$`), it's pretty good. an argument can be made that this proposal isn't needed as performance is already good enough. I think it would still be a nice to have though.

as for the type hint, couldn't it added implicitly the moment the caching happens?

**Comment by aaronfranke** [comment]
> as for the type hint, couldn't it added implicitly the moment the caching happens?

The caching happens at runtime, since the script can be used in multiple places and may have different children.

**Comment by dalexeev** [comment]
> alright I just read through the entire thread and while the critique made sense to me, I did not see a single comment arguing that `$$` was a bad idea.

There is the following comment and I agree with it:

> You guys know how I think about these things, I think that most of the time, _explicitness is better_, and things that hide what happens under the hood to the programmer in a non-obvious way, even if it means writing a bit less code, is generally bad.

If something is used frequently, then you need to store it somewhere (in a variable). This is a universal solution that works not only with `Node` references. `$$` is too narrow a solution, which is also confusing.

`$$` is purely for optimizations. A similar situation was with `iniline` (#3760). It seems to me that at the language level there should not be things intended only for optimizations.

> 278ms # get_node(^"LiteralNodePath")
> 533ms # get_node("NodePathAsString")

In theory, we can make the GDScript compiler smarter and optimize the `get_node(<constant string expression>)` case. This optimization will not add confusion.

**Comment by smeans** [comment]
I'm completely new to Godot, so I'm trying to decide if I want to be in the onready/get_node or $ camp. My first concern was the performance implication of the $ operator, so I found this thread. I understand not wanting to break backwards compatibility, but what if there were a source file or (preferably) project-level directive to enable caching (like "option explicit" in JavaScript)? It would let those who know what they're doing take advantage of caching right away without breaking old projects. Then, after a few releases the default value of the option could be flipped to cache by default.

**Comment by Calinou** [comment]
> I understand not wanting to break backwards compatibility, but what if there were a source file or (preferably) project-level directive to enable caching (like "option explicit" in JavaScript)?

We prefer not making GDScript behavior dependent on directives by this, or worse, project settings. Scripts may be copy-pasted from a project to another, and this should _never_ cause them to break because a project uses different settings from another.

**Comment by h0lley** [comment]
> EDIT: If https://github.com/godotengine/godot-proposals/issues/1048 is implemented, it may make this proposal obsolete.

I disagree that #1048 makes this proposal obsolete.

I thought this was at least in part about reducing the ugly boiler plate code of `onready` declarations. having to declare a bunch of `@export var` instead doesn't help with that.

> However, now this code is ugly, since it is much longer than it could be and we've lost the green syntax highlighting in new_game() that tells us we're working with node children.

**Comment by EchoingLogos** [comment]
An argument that I'd like to stress further in favor of this is in terms of code readability. Currently, using $ or % to access nodes provides the benefit to the reader of immediately knowing that the object being referenced is 1. A Node 2. Somewhere in the child hierarchy 3. How far down in the child hierarchy it is being directly proportional to the length of the path. This is a *ton* of information provided to the reader thanks to the use of $, % or get_node().

In the case of onready var, the user is now forced to cache the node into a run-of-the-mill variable indistinguishable at a glance from other global scope variables losing the previously described advantages, unless the user specifically breaks variable naming styles to obtain these advantages. This is run of the mill for programming, and there's "no reason" why I should expect one type of variable to be highlighted like this; but the way things are in Godot, I find myself using $ not because I don't want to cache the node, but simply because it makes the code far more readable.

I acknowledge that altering behavior for $ would break compatibility in very obscure ways amongst other issues, but I strongly support the implementation of $$ as a cached $.

**Comment by nlupugla** [comment]
I think Godot does have a genuine need for more static scene information, but I don't think this is the way to accomplish it.

The better way to solve the issues of node caching is actually with a change to the editor, not with GDScript. The idea would be the editor lets you make nodes in a scene as static. Then `$my_static_node` gains several benefits.

1. If the node doesn't exist in the scene tree, the code will highlighted in red at compile time.
2. The node will be typed.
3. The node will be cached.

Furthermore, this approach has no risk of breaking compatibility because all nodes would be marked as dynamic (ie: non-static) by default.

**Comment by aaronfranke** [comment]
@nlupugla Keep in mind that a script may be used for multiple scenes, so there is no way to find the static typing information ahead of time. Well, it can be done just in the script editor (it already does this for autocomplete hints, but doing it in GDScript in general at runtime is tricky.

@h0lley You are correct. To clarify, with exporting node references, that is a feature I personally would use over the `$` or `$$` operators, but cached `$`/`$$` would still be a nice feature to have. (I say "would use" because at the moment exporting node references is buggy so unfortunately I do not use it yet).

**Comment by nlupugla** [comment]
Fair point about a script being reused for multiple scenes. I think there are multiple good workarounds though. For example, there could be a declare at the top of a script what scene it will be statically attached to.

```
extends Node2D in "res://my_scene.tscn" on "%MyNode2D"
```

**Comment by dalexeev** [comment]
> For example, there could be a way to declare at the top of a script what scene it is statically attached to.

See #1935. But I'm still skeptical about static node validation. The node can still be deleted at runtime, its name and/or path can be changed.

**Comment by nlupugla** [comment]
Thanks for linking to that proposal dalexeev :)

In my mind, the whole point of marking a node as static within a scene would be that any attempts to delete, rename, or move it would fail at compile time. The node would be treated like a const.

**Comment by dalexeev** [comment]
> In my mind, the whole point of marking a node as static within a scene would be that any attempts to delete, rename, or move it would fail at compile time.

This is difficult or even impossible to guarantee. You can always cheat the static analyzer, sometimes unintentionally. Or the forbidden operation can be performed by external code or the engine.

```gdscript
var t: Variant = static_node
t.queue_free()
```

**Comment by nlupugla** [comment]
Fair enough. The program could abort upon trying to delete a static node though, if static analysis isn't enough to catch everything right?

By the way, I don't mean to digress too far from the original topic. On the subject of cached node references, my main problem with the idea is that it makes a bunch of implicit assumptions about the structure of your scene, whereas the static scene approach makes these assumptions explicit. In both approaches, there will be a problem if the node in question moves.

**Comment by nlupugla** [comment]
To be a bit more clear on the idea, I wrote up a proposal: https://github.com/godotengine/godot-proposals

**Comment by MadeScientist** [comment]
Personally, I adhere to the GDScript Style Guide and use an underscore prefix combined with PascalCase to distinguish node reference variables from regular variables. I'm not sure if others appreciate this style, but if it's possible to implement highlighting for _PascalCase variables in green, it would be quite good for me.

Like this:
![Image](https://github.com/user-attachments/assets/807bfa47-9cdc-4d87-b9f6-0863052eab7d)

**Comment by Zireael07** [comment]
This is about $ and/or caching, what does PascalCase have anything to do with it?

**Comment by MadeScientist** [comment]
> This is about $ and/or caching, what does PascalCase have anything to do with it?这是关于 $ 和/或缓存的，PascalCase 与它有什么关系？

Sorry for my misunderstanding.

**Comment by MadeScientist** [comment]
> This is about $ and/or caching, what does PascalCase have anything to do with it?

Sorry for my misunderstanding.

**Comment by EchoingLogos** [comment]
> I think Godot does have a genuine need for more static scene information, but I don't think this is the way to accomplish it.
> 
> The better way to solve the issues of node caching is actually with a change to the editor, not with GDScript. The idea would be the editor lets you mark nodes in a scene as static. Then `$my_static_node` gains several benefits.
> 
>     1. If the node doesn't exist in the scene tree, the code will highlighted in red at compile time.
> 
>     2. The node will be typed.
> 
>     3. The node will be cached.
> 
> 
> Furthermore, this approach has no risk of breaking compatibility because all nodes would be marked as dynamic (ie: non-static) by default.

I disagree. Specifically, referencing child nodes is a foundational aspect of writing GDScript, and its usability should be prioritized within the language itself rather than relegated to editor settings.

The $ operator often leads to performance pitfalls that are hard to debug, especially in complex projects. As it is usually difficult to determine exactly what is being called way too many times, developers are left with a choice: either forget about the lack of caching, or avoid $ and % altogether. The former is a huge performance trade-off for complex projects, and the latter is a huge readability trade-off for all projects. This undermines one of GDScript’s core advantages: its intuitive alignment with Godot's node-based architecture.

This is why I continue to advocate for introducing a $$ operator. It offers a natural evolution to $, prioritizing readability and path clarity while mitigating the performance drawbacks. While I understand that this proposal has not been super popular, I don’t believe the potential of the $$ operator has been thoroughly explored or addressed. I also invite @aaronfranke to consider the implementation of a $$ to be the main aspect of this proposal. 

Ultimately, GDScript is first and foremost, a language for the Godot engine, and the node hierarchy is central to how developers build with it. $$ has the potential to make a significant impact on usability and performance, and I’d welcome a deeper discussion on why this isn't as critical as it appears to me.
